runtime/tests/tests.rs

```diff
@@ -462,6 +518,125 @@ fn staking_validators_should_receive_equal_transaction_fee_reward() {
 		});
 }
 
+#[test]
+/// This tests if authorship reward of the last block in an era is factored in.
+fn authorship_reward_of_last_block_in_an_era() {
+	let validators_count = 6;
+	let initial_balance = 1_000 * TransactionBaseFee::get();
+
+	ExtBuilder::default()
+		.validator_count(validators_count)
+		.initial_balance(initial_balance)
+		.stash(initial_balance)
+		.build()
+		.execute_with(|| {
+			let last_session_of_era_0 = SessionsPerEra::get() - 1;
+			rotate_to_session(last_session_of_era_0);
+
+			// The last session falls in the era 0
+			assert_eq!(Staking::current_era(), 0);
+
+			// Make sure we have the correct number of validators elected
+			assert_eq!(Staking::current_elected().len(), validators_count);
+
+			// Make a block header whose author is specified as below
+			let author_index = 0; // index 0 of validators
+			let first_block_of_era_1 = System::block_number() + 1;
+			let header_of_last_block = header_for_block_number(first_block_of_era_1.into());
+			let header = specify_author(header_of_last_block, author_index.clone());
+
+			let author_stash_id = Session::validators()[(author_index as usize)].clone();
+
+			// The previous session should come to its end
+			pallet_babe::CurrentSlot::put(Babe::current_slot() + EpochDuration::get());
+
+			send_heartbeats();
+
+			let author_stash_balance_before_adding_block =
+				GenericAsset::free_balance(&SPENDING_ASSET_ID, &author_stash_id);
+
+			// Let's go through the first stage of executing the block
+			Executive::initialize_block(&header);
+
+			// initializing the last block should have advanced the session and thus changed the era
+			assert_eq!(Staking::current_era(), 1);
+
+			// No offences should happened. Thus the number of validators shouldn't have changed
+			assert_eq!(Staking::current_elected().len(), validators_count);
+
+			// There should be a reward calculated for the author
+			assert!(
+				GenericAsset::free_balance(&SPENDING_ASSET_ID, &author_stash_id)
+					> author_stash_balance_before_adding_block
+			);
+		});
+}
+
+#[test]
+/// This tests if authorship reward of the last block in an era is factored in, even when the author
+/// is chilled and thus not going to be an authority in the next era.
+fn authorship_reward_of_a_chilled_validator() {
+	let validators_count = 6;
+	let initial_balance = 1_000 * TransactionBaseFee::get();
+
+	ExtBuilder::default()
+		.validator_count(validators_count)
+		.initial_balance(initial_balance)
+		.stash(initial_balance)
+		.build()
+		.execute_with(|| {
+			let last_session_of_era_0 = SessionsPerEra::get() - 1;
+			rotate_to_session(last_session_of_era_0);
+
+			// The last session falls in the era 0
+			assert_eq!(Staking::current_era(), 0);
+
+			// make sure we have the correct number of validators elected
+			assert_eq!(Staking::current_elected().len(), validators_count);
+
+			// Make a block header whose author is specified as below
+			let author_index = 0; // index 0 of validators
```

Comment:
[suggestion] Extract this part of code in a func like `specify_header()` as it both use in this and above tests and have not dependency with other vars

Reply:
@HaoZhaiX I have had a second look at these few lines and realized they could be replaced with just one line like
`let header = specify_author(header_for_block_number(System::block_number() + 1), author_index.clone());`
So defining a new function for a one-liner is not really necessary. On the other side, I believe I have broken that one line into a few separate steps just for the sake of these two tests, so I can clarify how the header is built and why is it built this way.  So I prefer to keep them as they are. What do you think?